### ISSUE TITLE
expo-router: doesn't restore state when using browser back/forward buttons

### ISSUE THREAD

**Issue by kasrak** [expo/expo#31747]
### Minimal reproducible example

npx create-expo-app@latest; npm run web

### What platform(s) does this occur on?

Web

### Where did you reproduce the issue?

on web

### Summary

https://www.loom.com/share/feae2f6b331a4f3cb9951633b3f7cf70

This issue repros on web if you create the Expo starter project via npx create-expo-app@latest

Repro steps:
1. Load the app in a web browser
2. Click the "Explore" tab at the bottom
3. Click the browser's back button
4. Click the browser's forward button

Expected: you should be back on the Explore tab

Actual: you're still on the home page, and there's this error in the console:

```
The action 'RESET' with payload {"stale":false,"type":"stack","key":"stack-89f_vt6bznNwe8ZDmp0mR","index":0,"routeNames":["(tabs)","+not-found","_sitemap"],"routes":[{"name":"(tabs)","state":{"stale":false,"type":"tab","key":"tab-Ecfq_pT2uRroKfUj4Po-9","index":1,"routeNames":["index","explore"],"history":[{"type":"route","key":"index-dvBVUdqMOoqRmecNnOOhN"},{"type":"route","key":"explore-UpUnmnhhbnvqR2dlmlgQn"}],"routes":[{"name":"index","path":"/","key":"index-dvBVUdqMOoqRmecNnOOhN"},{"name":"explore","key":"explore-UpUnmnhhbnvqR2dlmlgQn"}]},"key":"(tabs)-qAmiwyZ8AwBcDAYrlGgYK"}]} was not handled by any navigator.
```

There is this issue from react-navigator which might have a clue on the fix: https://github.com/react-navigation/react-navigation/issues/9128

Someone in that thread said:

>  I was experiencing the same issue hitting back to the initial screen. My fix was to stop setting initialState on web as described in https://reactnavigation.org/docs/navigation-container/#initialstate

### Environment

```text
expo-env-info 1.2.0 environment info:
    System:
      OS: macOS 14.6.1
      Shell: 5.9 - /bin/zsh
    Binaries:
      Node: 20.10.0 - ~/.local/bin/node
      npm: 9.6.7 - ~/.nvm/versions/node/v18.17.1/bin/npm
      Watchman: 2024.05.06.00 - /opt/homebrew/bin/watchman
    Managers:
      CocoaPods: 1.15.2 - /usr/local/bin/pod
    SDKs:
      iOS SDK:
        Platforms: DriverKit 23.5, iOS 17.5, macOS 14.5, tvOS 17.5, visionOS 1.2, watchOS 10.5
    IDEs:
      Android Studio: 2024.1 AI-241.18034.62.2411.12071903
      Xcode: 15.4/15F31d - /usr/bin/xcodebuild
    npmPackages:
      expo: ~51.0.28 => 51.0.34
      expo-router: ~3.5.23 => 3.5.23
      react: 18.2.0 => 18.2.0
      react-dom: 18.2.0 => 18.2.0
      react-native: 0.74.5 => 0.74.5
      react-native-web: ~0.19.10 => 0.19.12
    Expo Workflow: managed
```

### Expo Doctor Diagnostics

```text
✔ Check Expo config for common issues
✔ Check package.json for common issues
✔ Check native tooling versions
✔ Check if the project meets version requirements for submission to app stores
✔ Check dependencies for packages that should not be installed directly
✔ Check for common project setup issues
✔ Check for app config fields that may not be synced in a non-CNG project
✔ Check npm/ yarn versions
✔ Check for issues with metro config
✔ Check for legacy global CLI installed locally
✔ Check that native modules do not use incompatible support packages
✔ Check Expo config (app.json/ app.config.js) schema
✔ Check that packages match versions required by installed Expo SDK
✔ Check that native modules use compatible support package versions for installed Expo SDK

Didn't find any issues with the project!
```

**Comment by dtrongphuc** [comment]
Same here
`expo-router`: 3.5.18 works fine

**Comment by linhvovan29546** [comment]
> Same here `expo-router`: 3.5.18 works fine

It worked when downgraded to 3.5.18

**Comment by cevatbarisyilmaz** [comment]
Having the same issue after creating a new project with `npx create-expo-app@latest`

**Comment by sensor** [comment]
> Same here `expo-router`: 3.5.18 works fine

I can confirm this. Works with 3.5.18. Did not work with 3.5.19.

**Comment by spehlivan** [comment]
> Same here `expo-router`: 3.5.18 works fine

Yes it works with 3.5.18 This is ridicilous :@

**Comment by Vimiso** [comment]
Any updates on this?

**Comment by stephentuso** [comment]
Still seeing this with expo 52 and expo-router `4.0.8`

**Comment by stephentuso** [comment]
This seems to have been introduced with the fork of [createMemoryHistory](https://github.com/expo/expo/pull/29990/files#diff-8e22bc1cd48d45e4074c0042c548120fde9c1d8af547e622f464579d9f0f3ef9) in `3.5.19`.

Switching back to `createMemoryHistory` from `react-navigation` seems to fix the issue, which I'm doing on `4.0.8` with the below patch. However, it looks like it was forked for something related to [pushing hashes](https://github.com/expo/expo/pull/29990), which I'm not using. I may have time to look into the hash issue and make a full PR next week.

expo-router+4.0.8.patch
```patch
diff --git a/node_modules/expo-router/build/fork/useLinking.js b/node_modules/expo-router/build/fork/useLinking.js
index 827867d..7b43d59 100644
--- a/node_modules/expo-router/build/fork/useLinking.js
+++ b/node_modules/expo-router/build/fork/useLinking.js
@@ -30,7 +30,7 @@ exports.useLinking = exports.series = void 0;
 const native_1 = require("@react-navigation/native");
 const fast_deep_equal_1 = __importDefault(require("fast-deep-equal"));
 const React = __importStar(require("react"));
-const createMemoryHistory_1 = require("./createMemoryHistory");
+const createMemoryHistory_1 = require('@react-navigation/native/lib/commonjs/createMemoryHistory');
 const serverLocationContext_1 = require("../global-state/serverLocationContext");
 /**
  * Find the matching navigation state that changed between 2 navigation states
```

**Comment by WKampel** [comment]
Can confirm that downgrading to Expo Router v3.5.17 fixed it for me.

**Comment by Zephilinox** [comment]
just spent the last 6 hours trying to debug this 😩 not fun as a new user, I thought I was doing something wrong.

I'm on expo router 4.0.9, downgrading to 3.5.17 with ` npm install --save expo-router@3.5.17 --legacy-peer-deps` fixed it

**Comment by serfgy** [comment]
> Can confirm that downgrading to Expo Router v3.5.17 fixed it for me.

confirm 3.5.17 works, going forward and backward retains state (as it should be)

**Comment by WKampel** [comment]
Downgrading is annoying because 3.5.18 doesn't type enforce my routes.

**Comment by stephentuso** [comment]
Working on e2e tests and a PR for this

**Comment by marklawlor** [comment]
We'll have this released soon. Its currently blocked by another issue that was published in `4.0.12`

**Comment by Vimiso** [comment]
Great news, however would it make sense to also patch version ~3? If at all possible? Leaving the latest version of 3 in a working state. Plus, many of us are still on SDK 51.

**Comment by WKampel** [comment]
> Great news, however would it make sense to also patch version ~3? If at all possible? Leaving the latest version of 3 in a working state. Plus, many of us are still on SDK 51.

This would be great.

**Comment by ericlyoung** [comment]
Spent the last few hours debugging this.  Glad to see I should just revert back to 3.5.18

**Comment by madebari** [comment]
is the problem solved in expo sdk 52 and "expo-router": "~4.0.14" ?. please confirm to me.

**Comment by stephentuso** [comment]
@madebari Yes it's published

**Comment by trcoffman** [comment]
Could you backport this to expo router v3?

**Comment by madmeatballs** [comment]
this issue is still in 4.0.17.

**Comment by VariabileAleatoria** [comment]
Please, 
since sdk52 has this open issue https://github.com/expo/expo/issues/34449 on navigation, a backport of this fix in router v3 would be much needed

**Comment by Zephilinox** [comment]
this is still broken for me on the latest version of expo-router 4.0.20, with expo 52.0.11, react 18.3.1, react-native 0.76.3, react-native-web 0.19.13:
 - start at `/foo/[foo_id]`, e.g `/foo/foo18274`
 - click link to navigate to `/foo/[foo_id]/bar/[bar_id]`, e.g `/foo/foo1827/bar/bar101`
 - click link to navigate to `/foo/foo_id]` (i.e going backwards via a `<Link href="/foo/foo18274">`
 - now try to go back via the browser (i.e mouse back button) and it will fail, keeping the user on `/foo/foo18274` instead of correctly going back to `/foo/foo1827/bar/bar101`

**Comment by Zephilinox** [comment]
hey @stephentuso sorry, just wanted to tag you in case you don't get notifications and might know more about this problem after your last fix

**Comment by stephentuso** [comment]
@Zephilinox sorry - I've moved away from expo router now so I don't have any more insight, I had a lot less issues just using react navigation directly

**Comment by shubhenduyednap** [comment]
hey @Zephilinox , did you find any solution for the problem, still facing the same problem with browser is refreshed or back button is pressed.

**Comment by Zephilinox** [comment]
> hey @Zephilinox , did you find any solution for the problem, still facing the same problem with browser is refreshed or back button is pressed.

nope :( sorry!

I was only building a side project so it wasn't a deal breaker, just frustrating. this was my first real project in the web domain and I have to say I had higher expectations about the quality and care that would be taken towards the user experience, so it's really disappointing that it not only hasn't been fixed, but also that the issue is still marked as solved

I've had so many other problems using these expo libraries that I just can't see myself making another project using them. If you have any other alternatives in mind then I suggest going for that instead.
